llvm/lib/SYCLLowerIR/ComputeModuleRuntimeInfo.cpp

```diff
@@ -354,6 +357,54 @@ PropSetRegTy computeModuleProperties(const Module &M,
     PropSet.add(PropSetRegTy::SYCL_MISC_PROP, "specConstsReplacedWithDefault",
                 1);
 
+  { // Properties related to virtual functions
+    StringSet<> UsedVFSets;
+    bool ProvidesVFSet = false;
+    for (const Function &F : M) {
+      if (F.isDeclaration())
+        continue;
+
+      if (F.hasFnAttribute("indirectly-callable")) {
+        PropSet.add(PropSetRegTy::SYCL_VIRTUAL_FUNCTIONS,
+                    "virtual-functions-set",
+                    F.getFnAttribute("indirectly-callable").getValueAsString());
+        ProvidesVFSet = true;
+        // Device code split should ensure that virtual functions that belong
+        // to different sets are split into separate device images and hence
+        // there is no need to scan other functions.
+        break;
+      }
+
+      if (F.hasFnAttribute("calls-indirectly")) {
+        assert(
+            !ProvidesVFSet &&
```

Comment:
We assume that device code split outlined all virtual functions into a separate device image already, so there couldn't be both function definitions with `calls-indirectly` and `indirectly-callable` attributes.

I think that this particular `assert` is also unnecessary, because we have a similar one when we add the second property and it should be enough to check for self-consistency

Reply:
ah right, i forgot this is run after split. thanks

Reply:
The `assert` should be less confusing now since 67df1f064eba6a1fe16244ad1793f9955fa4d452